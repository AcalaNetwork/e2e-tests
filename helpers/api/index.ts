import { ApiPromise } from '@polkadot/api'

export const xtokens = {
  relaychainV2: (acc: any) => ({
    V1: {
      parents: 1,
      interior: {
        X1: {
          AccountId32: {
            network: 'Any',
            id: acc,
          },
        },
      },
    },
  }),
  relaychainV3: (acc: any) => ({
    V3: {
      parents: 1,
      interior: {
        X1: {
          AccountId32: {
            id: acc,
          },
        },
      },
    },
  }),
  parachainV2: (paraId: number) => (acc: any) => ({
    V1: {
      parents: 1,
      interior: {
        X2: [
          { Parachain: paraId },
          {
            AccountId32: {
              network: 'Any',
              id: acc,
            },
          },
        ],
      },
    },
  }),
  parachainAccountId20V2: (paraId: number) => (acc: any) => ({
    V1: {
      parents: 1,
      interior: {
        X2: [
          { Parachain: paraId },
          {
            AccountKey20: {
              network: 'Any',
              key: acc,
            },
          },
        ],
      },
    },
  }),
  parachainV3: (paraId: number) => (acc: any) => ({
    V3: {
      parents: 1,
      interior: {
        X2: [
          { Parachain: paraId },
          {
            AccountId32: {
              network: 'Any',
              id: acc,
            },
          },
        ],
      },
    },
  }),
  transfer:
    (token: any, amount: any, dest: (dest: any) => any, weight: any = 'Unlimited') =>
    ({ api }: { api: ApiPromise }, acc: any) =>
      api.tx.xTokens.transfer(token, amount, dest(acc), weight),
  transferMulticurrencies:
    (token: any, amount: any, feeToken: any, feeAmount: any, dest: (dest: any) => any) =>
    ({ api }: { api: ApiPromise }, acc: any) =>
      api.tx.xTokens.transferMulticurrencies(
        [
          [token, amount],
          [feeToken, feeAmount],
        ],
        1,
        dest(acc),
        'Unlimited'
      ),
}

export const xcmPallet = {
  parachainV2: (parents: number, paraId: number) => ({
    V1: {
      parents,
      interior: {
        X1: { Parachain: paraId },
      },
    },
  }),
  relaychainV3: (acc: any) => ({
    V3: {
      parents: 1,
      interior: {
        X1: {
          AccountId32: {
            network: 'Any',
            id: acc,
          },
        },
      },
    },
  }),
  parachainV3: (parents: number, paraId: any) => ({
    V3: {
      parents,
      interior: {
        X1: { Parachain: paraId },
      },
    },
  }),
  limitedReserveTransferAssetsV2:
    (token: any, amount: any, dest: any) =>
    ({ api }: { api: ApiPromise }, acc: any) =>
      (api.tx.xcmPallet || api.tx.polkadotXcm).limitedReserveTransferAssets(
        dest,
        {
          V1: {
            parents: 0,
            interior: {
              X1: {
                AccountId32: {
                  network: 'Any',
                  id: acc,
                },
              },
            },
          },
        },
        {
          V1: [
            {
              id: token,
              fun: { Fungible: amount },
            },
          ],
        },
        0,
        'Unlimited'
      ),
  limitedReserveTransferAssetsV3:
    (token: any, amount: any, dest: any) =>
    ({ api }: { api: ApiPromise }, acc: any) =>
      (api.tx.xcmPallet || api.tx.polkadotXcm).limitedReserveTransferAssets(
        dest,
        {
          V3: {
            parents: 0,
            interior: {
              X1: {
                AccountId32: {
                  id: acc,
                },
              },
            },
          },
        },
        {
          V3: [
            {
              id: token,
              fun: { Fungible: amount },
            },
          ],
        },
        0,
        'Unlimited'
      ),
}

export const tx = {
  xtokens,
  xcmPallet,
}

export const query = {
  balances: ({ api }: { api: ApiPromise }, address: string) => api.query.system.account(address),
  tokens:
    (token: any) =>
    ({ api }: { api: ApiPromise }, address: string) =>
      api.query.tokens.accounts(address, token),
  assets:
    (token: number | bigint) =>
    ({ api }: { api: ApiPromise }, address: string) =>
      api.query.assets.account(token, address),
  evm:
    (contract: string, slot: string) =>
    ({ api }: { api: ApiPromise }, _address: string) =>
      api.query.evm.accountStorages(contract, slot),
}
