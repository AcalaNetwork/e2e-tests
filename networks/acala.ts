import { Config } from './types'

export type Vars = {
  relayToken: string
  relayLiquidToken: string
}

export default {
  polkadot: {
    name: 'acala' as const,
    endpoint: 'wss://acala-rpc-0.aca-api.network',
    relayToken: 'DOT',
    relayLiquidToken: 'LDOT',
  },
  kusama: {
    name: 'karura' as const,
    endpoint: 'wss://karura-rpc-0.aca-api.network',
    relayToken: 'KSM',
    relayLiquidToken: 'LKSM',
  },
  config: ({ alice, relayToken, relayLiquidToken }) => ({
    storages: {
      System: {
        account: [[[alice.address], { data: { free: 10 * 1e12 } }]],
      },
      Tokens: {
        accounts: [
          [[alice.address, { Token: relayToken }], { free: 10 * 1e12 }],
          [[alice.address, { Token: relayLiquidToken }], { free: 100 * 1e12 }],
        ],
      },
      Sudo: {
        key: alice.address,
      },
      Homa: {
        // avoid impact test outcome
        $removePrefix: ['redeemRequests', 'unbondings', 'toBondPool'],
        // so that bump era won't trigger unbond
        relayChainCurrentEra: '0x64000000',
      },
      EvmAccounts: {
        accounts: [[['0x82a258cb20e2adb4788153cd5eb5839615ece9a0'], alice.address]],
        evmAddresses: [[[alice.address], '0x82a258cb20e2adb4788153cd5eb5839615ece9a0']],
      },
      PolkadotXcm: {
        $removePrefix: ['versionNotifyTargets', 'versionNotifiers', 'supportedVersion'],
      },
    },
  }),
} satisfies Config<Vars>

export const acala = {
  paraId: 2000,
  paraAccount: '13YMK2eYoAvStnzReuxBjMrAvPXmmdsURwZvc62PrdXimbNy',
  dot: { Token: 'DOT' },
  dai: { Erc20: '0x54a37a01cd75b616d63e0ab665bffdb0143c52ae' },
  wbtc: { ForeignAsset: 5 },
} as const

export const karura = {
  paraId: 2000,
  paraAccount: '13YMK2eYoAvStnzReuxBjMrAvPXmmdsURwZvc62PrdXimbNy',
  ksm: { Token: 'KSM' },
  usdt: { ForeignAsset: 7 },
  dai: { Erc20: '0x4bb6afb5fa2b07a5d1c499e1c3ddb5a15e709a71' },
} as const
