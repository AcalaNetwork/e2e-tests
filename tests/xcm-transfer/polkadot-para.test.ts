import { Context } from '../../networks/types'
import { query, tx } from '../../helpers/api'

import { acala } from '../../networks/acala'
import { hydraDX } from '../../networks/hydraDX'
import { moonbeam } from '../../networks/moonbeam'
import { statemint } from '../../networks/statemint'

import buildTest from './shared'

const tests = [
  // statemint <-> acala
  {
    from: 'statemint',
    to: 'acala',
    name: 'WBTC',
    fromStorage: ({ alice }: Context) => ({
      System: {
        account: [[[acala.paraAccount], { data: { free: 10e10 } }]],
      },
      Assets: {
        account: [[[statemint.wbtcIndex, alice.address], { balance: 1e8 }]],
        asset: [[[statemint.wbtcIndex], { supply: 1e8 }]],
      },
    }),
    test: {
      xcmPalletHorizontal: {
        tx: tx.xcmPallet.limitedReserveTransferAssetsV2(statemint.wbtc, 1e7, tx.xcmPallet.parachainV2(1, acala.paraId)),
        fromBalance: query.assets(statemint.wbtcIndex),
        toBalance: query.tokens(acala.wbtc),
      },
    },
  },
  {
    // TODO: this failed with FailedToTransactAsset on statemint somehow
    from: 'acala',
    to: 'statemint',
    route: 'polkadot', // for sending DOT for fee
    name: 'WBTC',
    fromStorage: ({ alice }: Context) => ({
      Tokens: {
        Accounts: [[[alice.address, acala.wbtc], { free: 1e8 }]],
      },
    }),
    toStorage: ({ alice }: Context) => ({
      System: {
        account: [[[acala.paraAccount], { data: { free: 10e10 } }]],
      },
      Assets: {
        account: [
          [[statemint.wbtcIndex, acala.paraAccount], { balance: 10e8 }],
          [[statemint.wbtcIndex, alice.address], { balance: 10e8 }],
        ],
        asset: [[[statemint.wbtcIndex], { supply: 10e8 }]],
      },
    }),
    test: {
      xtokenstHorizontal: {
        tx: tx.xtokens.transferMulticurrencies(
          acala.wbtc,
          1e7,
          acala.dot, // fee
          1e10,
          tx.xtokens.parachainV2(statemint.paraId)
        ),
        fromBalance: query.tokens(acala.wbtc),
        toBalance: query.assets(statemint.wbtcIndex),
      },
    },
  },
  // acala <-> hydraDX
  {
    from: 'acala',
    to: 'hydraDX',
    name: 'DAI',
    fromStorage: {
      Evm: {
        accountStorages: [
          [
            [
              acala.dai.Erc20,
              '0x2aef47e62c966f0695d5af370ddc1bc7c56902063eee60853e2872fc0ff4f88c', // balanceOf(Alice)
            ],
            '0x0000000000000000000000000000000000000000000000056bc75e2d63100000', // 1e20
          ],
        ],
      },
    },
    test: {
      xtokenstHorizontal: {
        tx: tx.xtokens.transfer(acala.dai, 10n ** 18n, tx.xtokens.parachainV2(hydraDX.paraId)),
        fromBalance: query.evm(acala.dai.Erc20, '0x2aef47e62c966f0695d5af370ddc1bc7c56902063eee60853e2872fc0ff4f88c'),
        toBalance: query.tokens(hydraDX.dai),
      },
    },
  },
  {
    from: 'hydraDX',
    to: 'acala',
    name: 'DAI',
    fromStorage: ({ alice }: Context) => ({
      Tokens: {
        accounts: [[[alice.address, hydraDX.dai], { free: 10n * 10n ** 18n }]],
      },
    }),
    test: {
      xtokenstHorizontal: {
        tx: tx.xtokens.transfer(hydraDX.dai, 10n ** 18n, tx.xtokens.parachainV2(acala.paraId), 5e9),
        fromBalance: query.tokens(hydraDX.dai),
        toBalance: query.evm(acala.dai.Erc20, '0x2aef47e62c966f0695d5af370ddc1bc7c56902063eee60853e2872fc0ff4f88c'),
      },
    },
  },
  // acala <-> moonbeam
  {
    from: 'acala',
    to: 'moonbeam',
    route: 'polkadot',
    name: 'DOT',
    toAccount: ({ alith }: Context) => alith,
    test: {
      xtokenstHorizontal: {
        tx: tx.xtokens.transfer(acala.dot, 1e12, tx.xtokens.parachainAccountId20V2(moonbeam.paraId)),
        fromBalance: query.tokens(acala.dot),
        toBalance: query.assets(moonbeam.dot),
        checkUmp: true,
      },
    },
  },
  {
    from: 'moonbeam',
    to: 'acala',
    route: 'polkadot',
    name: 'DOT',
    fromAccount: ({ alith }: Context) => alith,
    fromStorage: ({ alith }: Context) => ({
      Assets: {
        account: [[[moonbeam.dot, alith.address], { balance: 10e12 }]],
      },
    }),
    test: {
      xtokenstHorizontal: {
        tx: tx.xtokens.transfer({ ForeignAsset: moonbeam.dot }, 1e12, tx.xtokens.parachainV2(acala.paraId)),
        fromBalance: query.assets(moonbeam.dot),
        toBalance: query.tokens(acala.dot),
        checkUmp: true,
      },
    },
  },
] as const

export type TestType = (typeof tests)[number]

buildTest(tests)
