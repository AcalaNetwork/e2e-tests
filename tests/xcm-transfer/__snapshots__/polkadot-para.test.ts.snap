// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`'acala' -> 'hydraDX' xcm transfer 'DAI' > xtokens transfer > balance on from chain 1`] = `99000000000000000000`;

exports[`'acala' -> 'hydraDX' xcm transfer 'DAI' > xtokens transfer > balance on to chain 1`] = `
{
  "free": "(rounded 100000000000000000000)",
  "frozen": 0,
  "reserved": 0,
}
`;

exports[`'acala' -> 'hydraDX' xcm transfer 'DAI' > xtokens transfer > from chain hrmp messages 1`] = `
[
  {
    "data": [
      "ConcatenatedVersionedXcm",
      {
        "v2": [
          {
            "reserveAssetDeposited": [
              {
                "fun": {
                  "fungible": "0x00000000000000000de0b6b3a7640000",
                },
                "id": {
                  "concrete": {
                    "interior": {
                      "x2": [
                        {
                          "parachain": 2000,
                        },
                        {
                          "generalKey": "0x0254a37a01cd75b616d63e0ab665bffdb0143c52ae",
                        },
                      ],
                    },
                    "parents": 1,
                  },
                },
              },
            ],
          },
          {
            "clearOrigin": null,
          },
          {
            "buyExecution": {
              "fees": {
                "fun": {
                  "fungible": "0x00000000000000000de0b6b3a7640000",
                },
                "id": {
                  "concrete": {
                    "interior": {
                      "x2": [
                        {
                          "parachain": 2000,
                        },
                        {
                          "generalKey": "0x0254a37a01cd75b616d63e0ab665bffdb0143c52ae",
                        },
                      ],
                    },
                    "parents": 1,
                  },
                },
              },
              "weightLimit": {
                "unlimited": null,
              },
            },
          },
          {
            "depositAsset": {
              "assets": {
                "wild": {
                  "all": null,
                },
              },
              "beneficiary": {
                "interior": {
                  "x1": {
                    "accountId32": {
                      "id": "0x88dc3417d5058ec4b4503e0c12ea1a0a89be200fe98922423d4334014fa6b0ee",
                      "network": {
                        "any": null,
                      },
                    },
                  },
                },
                "parents": 0,
              },
              "maxAssets": 1,
            },
          },
        ],
      },
    ],
    "recipient": 2034,
  },
]
`;

exports[`'acala' -> 'hydraDX' xcm transfer 'DAI' > xtokens transfer > to chain xcm events 1`] = `
[
  {
    "data": {
      "messageHash": "(hash)",
      "weight": {
        "proofSize": 0,
        "refTime": 400000000,
      },
    },
    "method": "Success",
    "section": "xcmpQueue",
  },
]
`;

exports[`'acala' -> 'hydraDX' xcm transfer 'DAI' > xtokens transfer > tx events 1`] = `
[
  {
    "data": {
      "assets": [
        {
          "fun": {
            "Fungible": 1000000000000000000,
          },
          "id": {
            "Concrete": {
              "interior": {
                "X2": [
                  {
                    "Parachain": 2000,
                  },
                  {
                    "GeneralKey": {
                      "data": "(hash)",
                      "length": 21,
                    },
                  },
                ],
              },
              "parents": 1,
            },
          },
        },
      ],
      "dest": {
        "interior": {
          "X2": [
            {
              "Parachain": "(rounded 2000)",
            },
            {
              "AccountId32": {
                "id": "(hash)",
                "network": null,
              },
            },
          ],
        },
        "parents": 1,
      },
      "fee": {
        "fun": {
          "Fungible": 1000000000000000000,
        },
        "id": {
          "Concrete": {
            "interior": {
              "X2": [
                {
                  "Parachain": 2000,
                },
                {
                  "GeneralKey": {
                    "data": "(hash)",
                    "length": 21,
                  },
                },
              ],
            },
            "parents": 1,
          },
        },
      },
      "sender": "23y3WetbNi6rDMgHmyRDjgpb7PnhgPotuPPawxruTMLYTLzG",
    },
    "method": "TransferredMultiAssets",
    "section": "xTokens",
  },
]
`;

exports[`'acala' -> 'moonbeam' xcm transfer 'DOT' > xtokens transfer > balance on from chain 1`] = `
{
  "free": 9000000000000,
  "frozen": 0,
  "reserved": 0,
}
`;

exports[`'acala' -> 'moonbeam' xcm transfer 'DOT' > xtokens transfer > balance on to chain 1`] = `
{
  "balance": "(rounded 1000000000000)",
  "extra": null,
  "reason": {
    "sufficient": null,
  },
  "status": "Liquid",
}
`;

exports[`'acala' -> 'moonbeam' xcm transfer 'DOT' > xtokens transfer > from chain ump messages 1`] = `
[
  {
    "v2": [
      {
        "withdrawAsset": [
          {
            "fun": {
              "fungible": 1000000000000,
            },
            "id": {
              "concrete": {
                "interior": {
                  "here": null,
                },
                "parents": 0,
              },
            },
          },
        ],
      },
      {
        "clearOrigin": null,
      },
      {
        "buyExecution": {
          "fees": {
            "fun": {
              "fungible": 500000000000,
            },
            "id": {
              "concrete": {
                "interior": {
                  "here": null,
                },
                "parents": 0,
              },
            },
          },
          "weightLimit": {
            "unlimited": null,
          },
        },
      },
      {
        "depositReserveAsset": {
          "assets": {
            "wild": {
              "all": null,
            },
          },
          "dest": {
            "interior": {
              "x1": {
                "parachain": 2004,
              },
            },
            "parents": 0,
          },
          "maxAssets": 1,
          "xcm": [
            {
              "buyExecution": {
                "fees": {
                  "fun": {
                    "fungible": 500000000000,
                  },
                  "id": {
                    "concrete": {
                      "interior": {
                        "here": null,
                      },
                      "parents": 1,
                    },
                  },
                },
                "weightLimit": {
                  "unlimited": null,
                },
              },
            },
            {
              "depositAsset": {
                "assets": {
                  "wild": {
                    "all": null,
                  },
                },
                "beneficiary": {
                  "interior": {
                    "x1": {
                      "accountKey20": {
                        "key": "0xf24ff3a9cf04c71dbc94d0b566f7a27b94566cac",
                        "network": {
                          "any": null,
                        },
                      },
                    },
                  },
                  "parents": 0,
                },
                "maxAssets": 1,
              },
            },
          ],
        },
      },
    ],
  },
]
`;

exports[`'acala' -> 'moonbeam' xcm transfer 'DOT' > xtokens transfer > to chain xcm events 1`] = `
[
  {
    "data": {
      "messageId": "(hash)",
      "outcome": {
        "Complete": {
          "proofSize": "(rounded 26000)",
          "refTime": "(rounded 690000000)",
        },
      },
    },
    "method": "ExecutedDownward",
    "section": "dmpQueue",
  },
]
`;

exports[`'acala' -> 'moonbeam' xcm transfer 'DOT' > xtokens transfer > tx events 1`] = `
[
  {
    "data": {
      "assets": [
        {
          "fun": {
            "Fungible": 1000000000000,
          },
          "id": {
            "Concrete": {
              "interior": "Here",
              "parents": 1,
            },
          },
        },
      ],
      "dest": {
        "interior": {
          "X2": [
            {
              "Parachain": "(rounded 2000)",
            },
            {
              "AccountKey20": {
                "key": "0xf24ff3a9cf04c71dbc94d0b566f7a27b94566cac",
                "network": null,
              },
            },
          ],
        },
        "parents": 1,
      },
      "fee": {
        "fun": {
          "Fungible": 1000000000000,
        },
        "id": {
          "Concrete": {
            "interior": "Here",
            "parents": 1,
          },
        },
      },
      "sender": "23y3WetbNi6rDMgHmyRDjgpb7PnhgPotuPPawxruTMLYTLzG",
    },
    "method": "TransferredMultiAssets",
    "section": "xTokens",
  },
]
`;

exports[`'acala' -> 'statemint' xcm transfer 'WBTC' > xtokens transfer > balance on from chain 1`] = `
{
  "free": 90000000,
  "frozen": 0,
  "reserved": 0,
}
`;

exports[`'acala' -> 'statemint' xcm transfer 'WBTC' > xtokens transfer > balance on to chain 1`] = `
{
  "balance": 1000000000,
  "extra": null,
  "reason": {
    "consumer": null,
  },
  "status": "Liquid",
}
`;

exports[`'acala' -> 'statemint' xcm transfer 'WBTC' > xtokens transfer > from chain hrmp messages 1`] = `
[
  {
    "data": [
      "ConcatenatedVersionedXcm",
      {
        "v2": [
          {
            "withdrawAsset": [
              {
                "fun": {
                  "fungible": 10000000,
                },
                "id": {
                  "concrete": {
                    "interior": {
                      "x2": [
                        {
                          "palletInstance": 50,
                        },
                        {
                          "generalIndex": 21,
                        },
                      ],
                    },
                    "parents": 0,
                  },
                },
              },
              {
                "fun": {
                  "fungible": 50000000,
                },
                "id": {
                  "concrete": {
                    "interior": {
                      "here": null,
                    },
                    "parents": 1,
                  },
                },
              },
            ],
          },
          {
            "clearOrigin": null,
          },
          {
            "buyExecution": {
              "fees": {
                "fun": {
                  "fungible": 50000000,
                },
                "id": {
                  "concrete": {
                    "interior": {
                      "here": null,
                    },
                    "parents": 1,
                  },
                },
              },
              "weightLimit": {
                "unlimited": null,
              },
            },
          },
          {
            "depositAsset": {
              "assets": {
                "wild": {
                  "all": null,
                },
              },
              "beneficiary": {
                "interior": {
                  "x1": {
                    "accountId32": {
                      "id": "0x88dc3417d5058ec4b4503e0c12ea1a0a89be200fe98922423d4334014fa6b0ee",
                      "network": {
                        "any": null,
                      },
                    },
                  },
                },
                "parents": 0,
              },
              "maxAssets": 2,
            },
          },
        ],
      },
    ],
    "recipient": 1000,
  },
]
`;

exports[`'acala' -> 'statemint' xcm transfer 'WBTC' > xtokens transfer > to chain xcm events 1`] = `
[
  {
    "data": {
      "messageId": "(hash)",
      "outcome": {
        "Complete": {
          "proofSize": 0,
          "refTime": "(rounded 160000000)",
        },
      },
    },
    "method": "ExecutedDownward",
    "section": "dmpQueue",
  },
  {
    "data": {
      "error": "FailedToTransactAsset",
      "messageHash": "(hash)",
      "weight": {
        "proofSize": "(rounded 7200)",
        "refTime": "(rounded 300000000)",
      },
    },
    "method": "Fail",
    "section": "xcmpQueue",
  },
]
`;

exports[`'acala' -> 'statemint' xcm transfer 'WBTC' > xtokens transfer > tx events 1`] = `
[
  {
    "data": {
      "assets": [
        {
          "fun": {
            "Fungible": 10000000000,
          },
          "id": {
            "Concrete": {
              "interior": "Here",
              "parents": 1,
            },
          },
        },
        {
          "fun": {
            "Fungible": 10000000,
          },
          "id": {
            "Concrete": {
              "interior": {
                "X3": [
                  {
                    "Parachain": 1000,
                  },
                  {
                    "PalletInstance": 50,
                  },
                  {
                    "GeneralIndex": 21,
                  },
                ],
              },
              "parents": 1,
            },
          },
        },
      ],
      "dest": {
        "interior": {
          "X2": [
            {
              "Parachain": 1000,
            },
            {
              "AccountId32": {
                "id": "(hash)",
                "network": null,
              },
            },
          ],
        },
        "parents": 1,
      },
      "fee": {
        "fun": {
          "Fungible": 10000000000,
        },
        "id": {
          "Concrete": {
            "interior": "Here",
            "parents": 1,
          },
        },
      },
      "sender": "23y3WetbNi6rDMgHmyRDjgpb7PnhgPotuPPawxruTMLYTLzG",
    },
    "method": "TransferredMultiAssets",
    "section": "xTokens",
  },
]
`;

exports[`'hydraDX' -> 'acala' xcm transfer 'DAI' > xtokens transfer > balance on from chain 1`] = `
{
  "free": 9000000000000000000,
  "frozen": 0,
  "reserved": 0,
}
`;

exports[`'hydraDX' -> 'acala' xcm transfer 'DAI' > xtokens transfer > balance on to chain 1`] = `"(rounded 999000000000000000)"`;

exports[`'hydraDX' -> 'acala' xcm transfer 'DAI' > xtokens transfer > from chain hrmp messages 1`] = `
[
  {
    "data": [
      "ConcatenatedVersionedXcm",
      {
        "v3": [
          {
            "withdrawAsset": [
              {
                "fun": {
                  "fungible": "0x00000000000000000de0b6b3a7640000",
                },
                "id": {
                  "concrete": {
                    "interior": {
                      "x1": {
                        "generalKey": {
                          "data": "0x0254a37a01cd75b616d63e0ab665bffdb0143c52ae0000000000000000000000",
                          "length": 21,
                        },
                      },
                    },
                    "parents": 0,
                  },
                },
              },
            ],
          },
          {
            "clearOrigin": null,
          },
          {
            "buyExecution": {
              "fees": {
                "fun": {
                  "fungible": "0x00000000000000000de0b6b3a7640000",
                },
                "id": {
                  "concrete": {
                    "interior": {
                      "x1": {
                        "generalKey": {
                          "data": "0x0254a37a01cd75b616d63e0ab665bffdb0143c52ae0000000000000000000000",
                          "length": 21,
                        },
                      },
                    },
                    "parents": 0,
                  },
                },
              },
              "weightLimit": {
                "unlimited": null,
              },
            },
          },
          {
            "depositAsset": {
              "assets": {
                "wild": {
                  "allCounted": 1,
                },
              },
              "beneficiary": {
                "interior": {
                  "x1": {
                    "accountId32": {
                      "id": "0x88dc3417d5058ec4b4503e0c12ea1a0a89be200fe98922423d4334014fa6b0ee",
                      "network": null,
                    },
                  },
                },
                "parents": 0,
              },
            },
          },
        ],
      },
    ],
    "recipient": 2000,
  },
]
`;

exports[`'hydraDX' -> 'acala' xcm transfer 'DAI' > xtokens transfer > to chain xcm events 1`] = `
[
  {
    "data": {
      "messageHash": "(hash)",
      "messageId": "(hash)",
      "weight": {
        "proofSize": 0,
        "refTime": 800000000,
      },
    },
    "method": "Success",
    "section": "xcmpQueue",
  },
]
`;

exports[`'hydraDX' -> 'acala' xcm transfer 'DAI' > xtokens transfer > tx events 1`] = `
[
  {
    "data": {
      "assets": [
        {
          "fun": {
            "Fungible": 1000000000000000000,
          },
          "id": {
            "Concrete": {
              "interior": {
                "X2": [
                  {
                    "Parachain": 2000,
                  },
                  {
                    "GeneralKey": {
                      "data": "(hash)",
                      "length": 21,
                    },
                  },
                ],
              },
              "parents": 1,
            },
          },
        },
      ],
      "dest": {
        "interior": {
          "X2": [
            {
              "Parachain": 2000,
            },
            {
              "AccountId32": {
                "id": "(hash)",
                "network": null,
              },
            },
          ],
        },
        "parents": 1,
      },
      "fee": {
        "fun": {
          "Fungible": 1000000000000000000,
        },
        "id": {
          "Concrete": {
            "interior": {
              "X2": [
                {
                  "Parachain": 2000,
                },
                {
                  "GeneralKey": {
                    "data": "(hash)",
                    "length": 21,
                  },
                },
              ],
            },
            "parents": 1,
          },
        },
      },
      "sender": "7Lh1DEaxk8TNVU7snC3UkSv7N7NHb7BUuEkYZBwrzfrPNQE3",
    },
    "method": "TransferredMultiAssets",
    "section": "xTokens",
  },
]
`;

exports[`'moonbeam' -> 'acala' xcm transfer 'DOT' > xtokens transfer > balance on from chain 1`] = `
{
  "balance": 9000000000000,
  "extra": null,
  "reason": {
    "consumer": null,
  },
  "status": "Liquid",
}
`;

exports[`'moonbeam' -> 'acala' xcm transfer 'DOT' > xtokens transfer > balance on to chain 1`] = `
{
  "free": "(rounded 11000000000000)",
  "frozen": 0,
  "reserved": 0,
}
`;

exports[`'moonbeam' -> 'acala' xcm transfer 'DOT' > xtokens transfer > from chain ump messages 1`] = `
[
  {
    "v3": [
      {
        "withdrawAsset": [
          {
            "fun": {
              "fungible": 1000000000000,
            },
            "id": {
              "concrete": {
                "interior": {
                  "here": null,
                },
                "parents": 0,
              },
            },
          },
        ],
      },
      {
        "clearOrigin": null,
      },
      {
        "buyExecution": {
          "fees": {
            "fun": {
              "fungible": 500000000000,
            },
            "id": {
              "concrete": {
                "interior": {
                  "here": null,
                },
                "parents": 0,
              },
            },
          },
          "weightLimit": {
            "unlimited": null,
          },
        },
      },
      {
        "depositReserveAsset": {
          "assets": {
            "wild": {
              "allCounted": 1,
            },
          },
          "dest": {
            "interior": {
              "x1": {
                "parachain": 2000,
              },
            },
            "parents": 0,
          },
          "xcm": [
            {
              "buyExecution": {
                "fees": {
                  "fun": {
                    "fungible": 500000000000,
                  },
                  "id": {
                    "concrete": {
                      "interior": {
                        "here": null,
                      },
                      "parents": 1,
                    },
                  },
                },
                "weightLimit": {
                  "unlimited": null,
                },
              },
            },
            {
              "depositAsset": {
                "assets": {
                  "wild": {
                    "allCounted": 1,
                  },
                },
                "beneficiary": {
                  "interior": {
                    "x1": {
                      "accountId32": {
                        "id": "0x88dc3417d5058ec4b4503e0c12ea1a0a89be200fe98922423d4334014fa6b0ee",
                        "network": null,
                      },
                    },
                  },
                  "parents": 0,
                },
              },
            },
          ],
        },
      },
    ],
  },
]
`;

exports[`'moonbeam' -> 'acala' xcm transfer 'DOT' > xtokens transfer > to chain xcm events 1`] = `
[
  {
    "data": {
      "messageHash": "(hash)",
      "messageId": "(hash)",
      "outcome": {
        "Complete": {
          "proofSize": 0,
          "refTime": 800000000,
        },
      },
    },
    "method": "ExecutedDownward",
    "section": "dmpQueue",
  },
]
`;

exports[`'moonbeam' -> 'acala' xcm transfer 'DOT' > xtokens transfer > tx events 1`] = `
[
  {
    "data": {
      "assets": [
        {
          "fun": {
            "Fungible": 1000000000000,
          },
          "id": {
            "Concrete": {
              "interior": "Here",
              "parents": 1,
            },
          },
        },
      ],
      "dest": {
        "interior": {
          "X2": [
            {
              "Parachain": 2000,
            },
            {
              "AccountId32": {
                "id": "(hash)",
                "network": null,
              },
            },
          ],
        },
        "parents": 1,
      },
      "fee": {
        "fun": {
          "Fungible": 1000000000000,
        },
        "id": {
          "Concrete": {
            "interior": "Here",
            "parents": 1,
          },
        },
      },
      "sender": "0xf24FF3a9CF04c71Dbc94D0b566f7A27B94566cac",
    },
    "method": "TransferredMultiAssets",
    "section": "xTokens",
  },
]
`;

exports[`'statemint' -> 'acala' xcm transfer 'WBTC' > xcmPallet transfer > balance on from chain 1`] = `
{
  "balance": 90000000,
  "extra": null,
  "reason": {
    "consumer": null,
  },
  "status": "Liquid",
}
`;

exports[`'statemint' -> 'acala' xcm transfer 'WBTC' > xcmPallet transfer > balance on to chain 1`] = `
{
  "free": "(rounded 10000000)",
  "frozen": 0,
  "reserved": 0,
}
`;

exports[`'statemint' -> 'acala' xcm transfer 'WBTC' > xcmPallet transfer > from chain hrmp messages 1`] = `
[
  {
    "data": [
      "ConcatenatedVersionedXcm",
      {
        "v3": [
          {
            "reserveAssetDeposited": [
              {
                "fun": {
                  "fungible": 10000000,
                },
                "id": {
                  "concrete": {
                    "interior": {
                      "x3": [
                        {
                          "parachain": 1000,
                        },
                        {
                          "palletInstance": 50,
                        },
                        {
                          "generalIndex": 21,
                        },
                      ],
                    },
                    "parents": 1,
                  },
                },
              },
            ],
          },
          {
            "clearOrigin": null,
          },
          {
            "buyExecution": {
              "fees": {
                "fun": {
                  "fungible": 10000000,
                },
                "id": {
                  "concrete": {
                    "interior": {
                      "x3": [
                        {
                          "parachain": 1000,
                        },
                        {
                          "palletInstance": 50,
                        },
                        {
                          "generalIndex": 21,
                        },
                      ],
                    },
                    "parents": 1,
                  },
                },
              },
              "weightLimit": {
                "unlimited": null,
              },
            },
          },
          {
            "depositAsset": {
              "assets": {
                "wild": {
                  "allCounted": 1,
                },
              },
              "beneficiary": {
                "interior": {
                  "x1": {
                    "accountId32": {
                      "id": "0x88dc3417d5058ec4b4503e0c12ea1a0a89be200fe98922423d4334014fa6b0ee",
                      "network": null,
                    },
                  },
                },
                "parents": 0,
              },
            },
          },
          {
            "setTopic": "redacted",
          },
        ],
      },
    ],
    "recipient": 2000,
  },
]
`;

exports[`'statemint' -> 'acala' xcm transfer 'WBTC' > xcmPallet transfer > to chain xcm events 1`] = `
[
  {
    "data": {
      "messageHash": "(hash)",
      "messageId": "(hash)",
      "weight": {
        "proofSize": 0,
        "refTime": 800000000,
      },
    },
    "method": "Success",
    "section": "xcmpQueue",
  },
]
`;

exports[`'statemint' -> 'acala' xcm transfer 'WBTC' > xcmPallet transfer > tx events 1`] = `
[
  {
    "data": [
      {
        "Complete": {
          "proofSize": "(rounded 6200)",
          "refTime": "(rounded 673000000)",
        },
      },
    ],
    "method": "Attempted",
    "section": "polkadotXcm",
  },
]
`;
